docs/customization_extendibility.md

```diff
@@ -0,0 +1,72 @@
+
+## Interacting with additional services
+Certain `ddev` commands, namely `ddev exec`, `ddev ssh`, and `ddev logs` interact with containers on an individual basis. By default, these commands interact with the web container for a project. All of these commands, however, provide a `--service` or `-s` flag allowing you to specify the service name of the container you want to interact with. For example, if you added a service to provide Apache Solr, and the service was named `solr`, you would be able to run `ddev logs --service solr` to retrieve the logs of the solr container.
+
+### Defining an additonal service - Apache Solr example
+The following is an example of an additional service that could be added to a project. This is a working example at time of writing. Annotations are added to highlight configurations important for running the service as part of a project in `ddev`. For full documentation on usage, see the [Docker Compose documentation](https://docs.docker.com/compose/overview/).
+
+This file would be saved as `docker-compose.solr.yml` in your project's `.ddev` folder.
+
+```
+# ddev apache solr recipe file
+#
+# To use this in your own project: Copy this file to your project's .ddev folder,
+# and create the folder path .ddev/solr/conf. Then, copy the solr configuration
+# files for your project to .ddev/solr/conf. E.g., using Drupal Search API Solr, 
+# you would copy the solr-conf/5.x/ contents into .ddev/solr/conf. The configuration
+# files must be present before running `ddev start`.
+
+version: '2'
+
+services:
+  solr: # This is the service name used when running ddev commands accepting the --service flag
+    container_name: local-${DDEV_SITENAME}-solr # This is the name of the container. It is recommended to follow the same name convention used in the main docker-compose.yml file.
+    image: solr:5.4
+    restart: always
+    ports:
+      - 8983 # Solr is served from this port inside the container
+    labels:
+    # These labels ensure this service is discoverable by ddev
+      com.ddev.site-name: ${DDEV_SITENAME}
+      com.ddev.approot: $DDEV_APPROOT
+      com.ddev.app-url: $DDEV_URL
+    environment:
+      - VIRTUAL_HOST=$DDEV_HOSTNAME # This defines the host name the service should be accessible from. This will be sitename.ddev.local
+      - VIRTUAL_PORT=8983 # This defines the port the service should be accessible from at sitename.ddev.local
+    volumes:
+      - "./solr:/solr-conf" # This exposes a mount to the host system `.ddev/solr-conf` directory.
+    entrypoint:
+      - docker-entrypoint.sh
+      - solr-precreate
+      - dev
+      - /solr-conf
+# This links the solr service to the web service defined in the main docker-compose.yml, allowing applications running in the web service to access the solr service at sitename.ddev.local:8983
+  web:
+    links:
+      - solr:$DDEV_HOSTNAME
+```
+
+## Overriding default container images
```

Comment:
This was tricky to cover without going into great detail, all of the details of which we don't have collected yet. I know _some_ of the things a replacement web or db container would need, but I doubt I would identify all of them without a deep dive. I'm hopeful this can be "enough for now" since this is an area we haven't focused much on, but we should consider how to surface the requirements for replacing these containers with user-provided alternatives.